qmail-qfilter.c

```diff
@@ -0,0 +1,330 @@
+/* This is free and unencumbered software released into the public * domain. */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <signal.h>
+#include <unistd.h>
+
+#include "qmail-qfilter.h"
+
+#ifndef TMPDIR
+#define TMPDIR "/tmp"
+#endif
+
+#ifndef BUFSIZE
+#define BUFSIZE 4096
+#endif
+
+#define QQ_OOM 51
+#define QQ_WRITE_ERROR 53
+#define QQ_INTERNAL 81
+#define QQ_BAD_ENV 91
+
+#define QQ_DROP_MSG 99
+
+#define MSGIN 0
+#define MSGOUT 1
+#define ENVIN 3
+#define ENVOUT 4
+#define QQFD 5
+
+static const char* qqargv[2];
+
+/* a replacement for setenv(3) for systems that don't have one */
+void mysetenv(const char* key, const char* val, size_t vallen)
+{
+  char* tmp;
+  size_t keylen;
+  
+  keylen = strlen(key);
+  tmp = malloc(keylen + 1 + vallen + 1);
+  memcpy(tmp, key, keylen);
+  tmp[keylen] = '=';
+  memcpy(tmp+keylen+1, val, vallen);
+  tmp[keylen+1+vallen] = 0;
+  if (putenv(tmp) != 0)
+    exit(QQ_OOM);
+}
+
+void mysetenvu(const char* key, unsigned long val)
+{
+  char buf[40];
+  int i;
+  i = sizeof buf;
+  buf[--i] = 0;
+  do {
+    buf[--i] = (val % 10) + '0';
+    val /= 10;
+  } while (val > 0);
+  mysetenv(key, buf + i, sizeof buf - 1 - i);
+}
+
+static size_t env_len = 0;
+static size_t msg_len = 0;
+
+/* Parse the sender address into user and host portions */
+size_t parse_sender(const char* env)
```

Comment:
This whole functions has a fundamental issue if the input is not "sane", i.e. if the data that is mmap'ed is not 0-terminated. In this case the ```strlen()``` below will run out of bounds and likely crash. The length passed to mmap will need to be passed to this function, and to ```parse_rcpts()``` as well so one can replace the ```strlen(env)``` by ```memchr(env, 0, envlen)```.

Reply:
I tried to address this upstream in https://github.com/bruceg/qmail-qfilter/pull/4